### ISSUE TITLE
Question about control mode and conversion

### ISSUE THREAD

**Issue by yeshenpy** [haosulab/ManiSkill#1146]
Thank you very much for your previous response. I would like to raise a few new questions related to an issue I encountered while training a reinforcement learning (RL) model.

First, does the choice of control mode make a significant difference? Specifically, I'm referring to the differences between joint qpos, end-effector (EE) delta position, and EE absolute position.

Intuitively, if we use joint qpos or EE absolute position as the action space, shouldn't we avoid constraining the action range to [-1, 1]?

Additionally, regarding data conversion, I noticed that converting joint qpos to EE delta position does not perfectly reproduce the original behavior. The resulting trajectory tends to be significantly longer, and there are also noticeable errors. Is there any solution to this issue?

**Comment by StoneT2000** [comment]
It is difficult to exactly reproduce behavior with a different controller all together. The mapping is not 1:1. Some reasons include the delta end effector controller can move faster/slower than the joint position ones simply depending on the current joint pos, just due to the nature of delta IK with a max limit on change in position (whereas joint control has max limit on change in joint angles). This is an open research problem in my opinion, or replay tool is just an attempt to provide one approach to conveying actions

Sometimes end effector control learning with RL can be faster wall time / sample efficiency. One reason is lower dimensionality of control. Another factor is sometimes EE control can permit smaller or larger max actions compared to joint control, which can affect the speed of RL training.

Moreover for RL training you should almost never use an absolute controller (unless you know the environment penalizes large actions as done in eg Unitree's isaac gym code for walking). You should almost always use delta control modes, RL learns far faster and is more stable. From a machine learning perspective this helps constrain data to a fixed normalized range and keeps the action distribution more uniform and avoids long tails.

**Comment by yeshenpy** [comment]
Thank you for your reply. I observed that using PD joint qpos control led to instability in the robot arm's learning process.

In addition, I did notice that videos converted using the ee delta mode appear more choppy and have longer durations compared to those using joint qpos.

We are currently converting based on two consecutive qpos values, which sometimes results in multiple ee delta steps. Conversely, is it possible to determine the maximum qpos change that can be achieved within a single ee delta step? In some cases, even discontinuous qpos transitions might be converted within one step. This raises the possibility of reconstructing the original qpos sequence by splitting it according to the step-wise limits of ee delta variations.

**Comment by yeshenpy** [comment]
Another question is whether there is a way to make the motion planning code complete the task within a specified number of steps. I’ve found that the provided code usually requires more environment steps to finish the task.

**Comment by StoneT2000** [comment]
You have to tune the motion planning code yourself then. Motion planning solutions do not have any whole trajectory optimization meaning each substage of the task has the robot stopping at the end of the stage, compared to more smooth movement achieved by RL methods that optimize the whole problem. 

You can probably increase the joint velocity/acceleration limits that mplib uses but that will only work to an extent as going too fast loses precision.

**Comment by yeshenpy** [comment]
I have a few new questions. After loading an articulated object, is it still possible to use the same function to check whether the object is being grasped—for instance, to determine whether the handle of a drawer is currently grasped? If this is not feasible, what alternative solutions would you recommend?

**Comment by StoneT2000** [comment]
I believe it is possible. is_grasped checks pairwise contact forces between the grippers and a desired actor or link. You should be able to just put in the cabinet link (or if its the whole drawer the drawer itself in case the handle is not separated from the drawer).

**Comment by yeshenpy** [comment]
Thanks! I noticed a minor issue: for the cabinet asset loaded from partnet, the door automatically opens with just a slight touch from the robotic arm, without requiring any pulling action. What could be the reason for this? Is the friction coefficient too low?

> 
        loader = self.scene.create_urdf_loader()
        density = 10
        static_friction = 1.0
        dynamic_friction = 1.0
        restitution = 0.0
        loader.fix_root_link = True  # or False
        # change the scale of the loaded articulation geometries (visual+collision)
        # if collision meshes contain multiple convex meshes
        # you can set this to True to try and load them
        loader.load_multiple_collisions_from_file = True
        # set density of all links
        loader.set_density(density)

        loader.set_material(static_friction, dynamic_friction, restitution)
        id = 49038
        with open(f'bounding_box.json', 'r') as f:
            bbox = json.load(f)
        
        min_pt = bbox['min']
        max_pt = bbox['max']
        
        # Compute dimensions
        dims = [max_pt[i] - min_pt[i] for i in range(3)]
        
        true_length = 0.5
        
        scale = true_length/dims[0]
        loader.scale = scale  # default is 1.0
   
        urdf_path = f"/dataset/{id}/mobility.urdf"
        articulation_builders = loader.parse(str(urdf_path))["articulation_builders"]
        builder = articulation_builders[0]
        #builder.initial_pose = sapien.Pose(p=[0, 0, 100000.0])
        self.cabinet= builder.build(name="cabinet")

**Comment by yeshenpy** [comment]
Oh, while manipulating the cabinet, I encountered an additional issue: I’m unable to access the drawer handle as a separate actor. After inspecting the mobile cabinet files, it seems that the handle is embedded as a mesh within a link, making it difficult to determine whether the robotic arm has actually grasped it.

**Comment by yeshenpy** [comment]
I’ve found that judging the success of grasps involving articulated objects is especially challenging, e.g., handle.
Are there any suggestions or ManiSkill features I might be missing?

**Comment by StoneT2000** [comment]
One trick is to not use contact forces and to instead check distance of each finger tip to the handle. You can see how we sample the handle mesh alone separate if the drawer link in the open cabinet drawer code.

**Comment by yeshenpy** [comment]
I noticed that every time I register a scene, I have to call remove_from_state_dict_registry followed by add_to_state_dict_registry.

1. I couldn’t find any relevant documentation — could you clarify the purpose of these two functions? Are they used for state construction or something related?

2. Another question: I noticed that in the cabinet drawer task, friction doesn’t seem to be configured — I couldn’t find the relevant settings in the associated configuration files. Could you provide any recommendations or best practices for setting an appropriate friction value?

3. Additionally, I noticed that the current version doesn’t seem to support setting the joint qpos.  self.handle_link.joint.qpos = 0.1 will lead to an attribute set error. [https://maniskill.readthedocs.io/en/latest/_modules/mani_skill/utils/structs/articulation_joint.html](https://maniskill.readthedocs.io/en/latest/_modules/mani_skill/utils/structs/articulation_joint.html)

4. One last question: In motion planning, sometimes certain joints end up touching the table, causing the robot to get stuck. Is there any way to avoid this?

**Comment by StoneT2000** [comment]
1. https://maniskill.readthedocs.io/en/latest/user_guide/tutorials/custom_tasks/advanced.html#handling-heterogeneous-simulation-states its on the documentation, you can look up the function to find some helpful explanations about what they do. I personally don't like the fact I had to add a new function for people to learn to enable some behavior but it was better than making assumptions of how users want to manage sim state
2. There arne't any best practices for friction. At least for sim2real you probably want to randomize the joint friction values
3. Added that feature just now via https://github.com/haosulab/ManiSkill/pull/1162
4. You need to tune your motion planning code to avoid the table. Hardcode some constraints / dont make it move too low.
